### ISSUE TITLE
`loadstring`: why do sstrict violations fail more severely than invalid code? 

### ISSUE THREAD

**Issue by bjornbm** [2dengine/sstrict.lua#2]
When loading invalid code with plain `loadstring` we get something like:

```
> print(loadstring [['a' + 1]])
nil     [string "'a' + 1"]:1: unexpected symbol near ''a''
```

That is, a falsy first value and the error message. In effect, loadstring behaves a bit like `pcall` in that it doesn't error on bad inputs.

When sstrict detects an error, on the other hand, an error is thrown:

```
> require 'sstrict'
> loadstring[[return a]]
./sstrict.lua:902:
return a:1: undefined variable 'a'
stack traceback:
        [C]: in function 'error'
        ./sstrict.lua:902: in function 'error'
        ./sstrict.lua:890: in function 'access'
        ./sstrict.lua:246: in function 'prefixexp'
        ./sstrict.lua:320: in function 'term'
        ./sstrict.lua:337: in function 'unaryexp'
        ./sstrict.lua:362: in function 'muldivexp'
        ./sstrict.lua:372: in function 'addsubexp'
        ./sstrict.lua:382: in function 'concatexp'
        ./sstrict.lua:392: in function 'relational'
        ./sstrict.lua:402: in function 'logical'
        ./sstrict.lua:412: in function 'expression'
        ./sstrict.lua:424: in function 'explist'
        ./sstrict.lua:714: in function 'chunk'
        ./sstrict.lua:727: in function 'block'
        ./sstrict.lua:759: in function 'reset'
        ./sstrict.lua:963: in function 'parse'
        ./sstrict.lua:981: in function 'loadstring'
        stdin:1: in main chunk
        [C]: at 0x0104c724bc
```

We can of course use `pcall` (as is done in `validate.lua`):

```
> print(pcall(loadstring, [[return a]]))
false   ./sstrict.lua:902:
return a:1: undefined variable 'a'
```

But this is a bit cumbersome and means that invalid code needs to be unwrapped one level: the falsy value from the failures is becomes the second return value behind a truthy value (since `loadstring` did not throw an error):

```
> print(pcall(loadstring, [['a' + 1]]))
true    nil     [string "'a' + 1"]:1: unexpected symbol near ''a''
```

Which brings me to my question: why don't sstrict violations fail in the same way as invalid code in `loadstring`? Is there a real interest to have them fail more severely (and require `pcall` to recover) than invalid code?

Thanks! And thanks for a cool project!

**Comment by 2dengine** [comment]
Hello. This is a good question.
I have to think about it some more before I can come up with a good answer.
The current behavior of the library is based merely on my own experience.

Here is what I have found so far:
1. sstrict.lua is slow so don't run it in production code
2. we want to catch mistakes as early as possible during development
3. it is good to be "loud" when you encounter mistakes during development

If you want to load a chunk without throwing errors you can use:
loadstring("--!strict\n"..code)

If you want to forward all errors to the console output you can use:
require("sstrict").panic = false

There is also --!sstrict which ignores all errors without spamming the console output.

**Comment by bjornbm** [comment]
Thanks for your observations! And for sharing the `panic = false` trick.

My feeling is anyway that it would make sense to align with the behaviors for invalid code in `load`, `loadfile`, and `dofile`:

- `loadstring`: has “protected” behavior (similar to `pcall`),
- `loadfile`: similar to `loadstring`
- `dofile`: “[In case of errors, `dofile` propagates the error to its caller (that is, `dofile` does not run in protected mode).](https://www.lua.org/manual/5.1/manual.html#pdf-dofile)” 

 Currently the corresponding functions provided by sstrict all behave the same with respect to sstrict violations: they all error to the caller.

In short, I would propose to wrap the `api.parse` and `api.parseFile` in `api.loadstring` and `api.loadpath` (but not in `api.dofile` in `pcall`s. Like so:

```lua
function api.loadstring(source, ...)
  local res, err = pcall(api.parse, source, ...)
  if not res then return res, err end
  return _loadstring(source, ...)
 end
```

I leave you to think about it. Let me know in the end if you want a pull request for this (which will also imply updating `try` in `validate.lua`).

**Comment by 2dengine** [comment]
Sure, you are welcome to send in a pull request if you want to change the public repository.

**Comment by bjornbm** [comment]
Hi, I've been working on this a bit today (had a busy week) but see you also got busy today. I'll try to reassess and (if necessary) make clean commits on top of your main branch.

**Comment by 2dengine** [comment]
I have already changed the behavior of api.loadstring as requested.
There are some known issues with api.require, but the library should work fine in most environments.
